### PR TITLE
feat(libflux): added support for all basic types in string interpolation

### PR DESCRIPTION
**Pull request by Bolladeen**
The following PR allows for all basic type (Bool, Int, Uint, Float, Time, Duration) to be acceptable string interpolation parts in the flux language. 

### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

File: compiler/compiler.go [hunk 1]
```diff
@@ -546,9 +546,14 @@ func compile(n semantic.Node, subst map[uint64]semantic.MonoType, scope Scope) (
 		if err != nil {
 			return nil, err
 		}
-		return &interpolatedEvaluator{
+		val := &interpolatedEvaluator{
```
Comment (rockstar) [inline compiler/compiler.go:549]: Is `val` an `evaluator` or a `value`? Is there a better, more detailed name we could give this variable?

File: libflux/core/src/semantic/nodes.rs [hunk 2]
```diff
@@ -608,11 +608,12 @@ impl StringExpr {
...
                 let (e, cons) = ip.expression.infer(env, f)?;
                 constraints.append(&mut Vec::from(cons));
-                constraints.push(Constraint::Equal {
-                    exp: MonoType::String,
+                constraints.push(Constraint::Kind {
+                    exp: Kind::Stringable,
                     act: ip.expression.type_of(),
                     loc: ip.expression.loc().clone(),
                 });
+                //TODO: add Kind::Stringable constraint
```
Comment (rockstar) [inline libflux/core/src/semantic/nodes.rs:616]: Is there an open issue for this?

**Review by rockstar — APPROVED** [review]
This looks good, thanks for doing it. One seemingly minor change requested, but happy to see this merge.

File: libflux/core/src/semantic/nodes.rs (hunk 2 above)
Comment (Bolladeen) [reply libflux/core/src/semantic/nodes.rs:616]: Good catch, it was a mental note I was making and should be deleted.

File: compiler/compiler_test.go [hunk 3]
```diff
@@ -44,12 +44,12 @@ func TestCompileAndEval(t *testing.T) {
 			fn:   `(r) => "n = ${r.n}"`,
```
Comment (nathanielc) [inline compiler/compiler_test.go:44]: Can you add a test case that shows interpolation working for other Stringable types besides string?

File: docs/SPEC.md [hunk 4]
```diff
@@ -325,6 +325,15 @@ Interpolation example:
 
 [IMPL#1775](https://github.com/influxdata/flux/issues/1775) Interpolate arbitrary expressions in string literals
 
+String interpolation now supports all basic type arbitrary expressions  without requiring casting to a string.
```
Comment (nathanielc) [inline docs/SPEC.md:328]: nitpit, but can you remove the `now` wording? We want the SPEC to read in the present tense as people reading the SPEC won't necessarily know or remember the history of Flux. 

In the SPEC we list the type classes can you add Stringable to that list?

File: interpreter/interpreter.go [hunk 5]
```diff
@@ -496,6 +496,11 @@ func (itrp *Interpreter) doStringPart(ctx context.Context, part semantic.StringE
 			return nil, errors.Newf(codes.Invalid, "%s: interpolated expression produced a null value",
 				p.Location())
 		} else if v.Type().Nature() != semantic.String {
+			val := values.NewString(fmt.Sprintf("%v", values.Unwrap(v)))
```
Comment (nathanielc) [inline interpreter/interpreter.go:499]: Do we want to rely on Go's formatting of the value here? Or should we be more explicit about how each of the Stringable types is converted to a string? For example using `strconv.FormatInt` etc?

File: interpreter/interpreter_test.go [hunk 6]
```diff
@@ -61,11 +61,13 @@ func TestEval(t *testing.T) {
 			wantErr: any,
 		},
 		{
-			name: "string interpolation field has wrong type",
+			name: "string interpolation non-string type",
 			query: `
 				r = makeRecord(o: {a: "foo", b: 42})
 				"r._value = ${r.b}"`,
```
Comment (nathanielc) [inline interpreter/interpreter_test.go:67]: Can you update this to have both cases? A case where it errors and a case where it succeeds? For example trying to format `r` in the string would still error, `"r = ${r}"`

**Review by nathanielc — COMMENTED** [review]
Looks really good, I'd like to see a few more tests. I called out the specific ones below. Other than that this looks ready to go!

File: docs/SPEC.md (hunk 4 above)
Comment (Bolladeen) [reply docs/SPEC.md:328]: On a related note, should I talk to docs team about the updated functionality of string interpolation?

File: interpreter/interpreter.go (hunk 5 above)
Comment (Bolladeen) [reply interpreter/interpreter.go:499]: I think either can work. As far as I could tell when testing, the actual output didn't differ for the basic types. I can change it just in case that behavior changes in the future.

File: docs/SPEC.md (hunk 4 above)
Comment (nathanielc) [reply docs/SPEC.md:328]: Yes, ping them in their channel with a link to this PR

File: interpreter/interpreter.go (hunk 5 above)
Comment (nathanielc) [reply interpreter/interpreter.go:499]: I think we should be explicit about how we are formatting the values. I don't expect Go's definition to change, but it would be best to be explicit so we are not tied to Go's behaviors.

File: compiler/compiler_test.go [hunk 7]
```diff
@@ -49,7 +49,37 @@ func TestCompileAndEval(t *testing.T) {
...
+			name: "interpolated string expression with duration type",
+			fn:   `(r) => "n = ${r.n}"`,
+			inType: semantic.NewObjectType([]semantic.PropertyType{
+				{Key: []byte("r"), Value: semantic.NewObjectType([]semantic.PropertyType{
+					{Key: []byte("n"), Value: semantic.BasicDuration},
+				})},
+			}),
+			input: values.NewObjectWithValues(map[string]values.Value{
+				"r": values.NewObjectWithValues(map[string]values.Value{
+					"n": values.NewString("1m"),
```
Comment (nathanielc) [inline compiler/compiler_test.go:67]: Why is this is a `values.NewString`? I expected it to be a  `values.NewDuration`

File: docs/SPEC.md [hunk 8]
```diff
@@ -325,6 +325,15 @@ Interpolation example:
 
 [IMPL#1775](https://github.com/influxdata/flux/issues/1775) Interpolate arbitrary expressions in string literals
```
Comment (nathanielc) [inline docs/SPEC.md:326]: This line can be removed with this change as it now implements arbitrary interpolation.

File: docs/SPEC.md [hunk 9]
```diff
@@ -325,6 +325,15 @@ Interpolation example:
...
 
+String interpolation supports all basic type arbitrary expressions  without requiring casting to a string. 
+Basic types include integer literals, unsigned integer literals, float literals, boolean literals, time literals and duration literals.
```
Comment (nathanielc) [inline docs/SPEC.md:329]: another nit (I am picky about the wording in the SPEC :D ) I prefer to use only concepts that have been defined in the SPEC so that it can be a self contained document as much as possible. We haven't defined what `casting` or `basic types` are so my ask is reword this to use the `Stringable` type constraint since that is now defined below. Something like:

String interpolation expressions must satisfy the Stringable constraint.

File: interpreter/interpreter.go [hunk 10]
```diff
@@ -496,6 +497,21 @@ func (itrp *Interpreter) doStringPart(ctx context.Context, part semantic.StringE
...
+				return values.NewString(strconv.FormatInt(val.(int64), 10)), nil
+			case semantic.UInt:
+				return values.NewString(strconv.FormatUint(val.(uint64), 10)), nil
+			case semantic.Float:
+				return values.NewString(strconv.FormatFloat(val.(float64), 'f', -1, 64)), nil
+			case semantic.Time:
+				return values.NewString(val.(values.Time).String()), nil
+			case semantic.Duration:
+				return values.NewString(val.(values.Duration).String()), nil
+			}
```
Comment (nathanielc) [inline interpreter/interpreter.go:514]: Where is the equivalent logic for the compiler? Maybe they can share the same implementation? This logic only uses types from the `values` package, perhaps a helper method in the values package like `Stringify(v Value) (String, error)` would be useful?

**Review by nathanielc — APPROVED** [review]
